core/src/main/java/org/apache/brooklyn/util/core/crypto/SecureKeys.java

```diff
@@ -78,11 +78,11 @@ public static X509Principal getX509PrincipalWithCommonName(String commonName) {
         return new X509Principal("" + "C=None," + "L=None," + "O=None," + "OU=None," + "CN=" + commonName);
     }
 
-    /** reads RSA or DSA / pem style private key files (viz {@link #toPem(KeyPair)}), extracting also the public key if possible
+    /** reads RSA or DSA / pem style private key files (viz {@link #toPem(KeyPair)}), extracting also the public key if possible. Closes the stream.
      * @throws IllegalStateException on errors, in particular {@link PassphraseProblem} if that is the problem */
     public static KeyPair readPem(InputStream input, final String passphrase) {
         // TODO cache is only for fallback "reader" strategy (2015-01); delete when Parser confirmed working
-        byte[] cache = Streams.readFully(input);
+        byte[] cache = Streams.readFullyAndClose(input);
```

Comment:
Feels like the responsibility of the caller to close the stream, rather than this method.

Looking at things like Guava's `ByteStreams`, they've deprecated all the methods that closed the stream that was passed in (and switched to a completely different pattern that uses `ByteSource`).

That makes me think it is the wrong thing to close the stream we are given.

Are there any good examples of good modern APIs that close the stream that is passed in?